vochain/state_snapshot.go

```diff
@@ -0,0 +1,416 @@
+	defer s.lock.Unlock()
+
+	// check if no more trees available
+	if s.currentTree >= len(s.header.Trees)-1 {
+		return io.EOF
+	}
+	s.currentTree++
+
+	// move the pointer is in the start of the next tree
+	seekPointer := snapshotHeaderLenSize + int(s.headerSize)
+	for i := 0; i < s.currentTree; i++ {
+		seekPointer += int(s.header.Trees[i].Size)
+	}
+	_, err := s.file.Seek(int64(seekPointer), io.SeekStart)
+
+	// update the buffer Reader
+	s.currentTreeReadBuffer = io.LimitReader(
+		s.file,
+		int64(s.header.Trees[s.currentTree].Size),
+	)
+	return err
+}
+
+// Read implements the io.Reader interface. Returns io.EOF error when no
+// more bytes available in the current three.
+func (s *StateSnapshot) Read(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	return s.currentTreeReadBuffer.Read(b)
+}
+
+// ReadAll reads the full content of the current tree and returns its bytes.
+// io.EOF error is returned if the bytes have been already read.
+func (s *StateSnapshot) ReadAll() ([]byte, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	b, err := io.ReadAll(s.currentTreeReadBuffer)
+	if err != nil {
+		return nil, err
+	}
+	if len(b) == 0 {
+		return nil, io.EOF
+	}
+	return b, nil
+}
+
+// Header returns the header for the snapshot containing the information
+// about all the merkle trees.
+func (s *StateSnapshot) Header() *SnapshotHeader {
+	return &s.header
+}
+
+// TreeHeader returns the header for the current tree.
+func (s *StateSnapshot) TreeHeader() *SnapshotHeaderTree {
+	return &s.header.Trees[s.currentTree]
+}
+
+// Path returns the file path of the snapshot file currently used.
+func (s *StateSnapshot) Path() string {
+	return s.path
+}
+
+// Create starts the creation of a new snapshot as a disk file.
+// This method must be called only once and its operation is oposed to `Open`.
+func (s *StateSnapshot) Create(filePath string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	var err error
+	s.header.Version = snapshotHeaderVersion
+	s.file, err = os.Create(filePath + ".tmp")
+	s.path = filePath
+	return err
+}
+
+// AddTree adds a new tree to the snapshot. `Create` needs to be called first.
+func (s *StateSnapshot) AddTree(name, parent string, root []byte) {
+	s.lock.Lock() // only 1 tree at time is allowed
+	s.header.Trees = append(s.header.Trees, SnapshotHeaderTree{
+		Name:   name,
+		Parent: parent,
+		Root:   root,
+		Size:   0,
+	})
+}
+
+// EndTree finishes the addition of a tree. This method should be called after `AddTree`.
+func (s *StateSnapshot) EndTree() {
+	s.currentTree++
+	s.lock.Unlock()
+}
+
+// Save builds the snapshot started with `Create` and stores in disk its contents.
+// After calling this method the snapshot is finished.
+// `EndTree` must be called before saving.
+func (s *StateSnapshot) Save() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	// create the final file
+	finalFile, err := os.Create(s.path)
+	defer func() {
+		finalFile.Close()
+		if err != nil {
+			log.Error(err)
+		}
```

Comment:
this is weird. you're not actually checking the close error? also, you should return the error, not log it. so this should go at the end of the function instead, and here you should do a `defer finalFile.Close()`.

Reply:
right